### ISSUE TITLE
Implement Index API

### ISSUE THREAD

**Issue by markmeeus** [markmeeus/MarcelloDB#28]
I'm using this issue to write down some ideas about the index api.

**Comment by markmeeus** [comment]
a few idea's

``` c#
collection.createIndex("Name", p => p.Name);
collection.createIndex("FirstName_Name", p => new MarcelloDB.Compound(p.FirstName, p.Name));
//...
index = collection.index["Name"];
index = collection.index["FirstName_Name"];
```

or

``` c#
collection.initializeIndexes((indexes)=>{
  indexes.create("Name", p => p.Name);
  indexes.create("FirstName_Name", new MarcelloDB.Compound(p.FirstName, p.Name));
});
//...
index = collection.index["name"];
index = collection.index["FirstName_Name"];
```

or 

``` c#
class MyIndexes : MarcelloDB.CollectionIndexes<Product>{
  public string Name(Product p){
    return p.name;
  }
  public MarcelloDB.Compound FirstName_Name(Product p){
    new MarcelloDB.Compound(p.firstName, p.name)
  }
}
collection.Index<MyIndexes>();
//...
index = collection.Index<MyIndex>.IndexFor(idx => idx.Name);
index = collection.Index<MyIndex>.IndexFor(idx => idx.FirstName_Name);
```

First one is easiest to remember and least surprise.
Second one has the benefit of beeing a single call, this allows us to cleanup unused indexes.

Third one eliminates the string to name the index, which will give better compiler support.
With the last option it is not possible to reference a non-existing index without the compiler noticing it.

**Comment by markmeeus** [comment]
Since the indexing mechanism will be such an important part of the api, and since the class based index definition seems a good idea atm, why not use it as a generic type of the collection itself.

``` c#
class ProductIndex{
  public string Name(Product p){
   return p.Name
  }
  public int Type(Product p){
    return p.ProductType; //name of the index 'Type' is not related to name of the property 'ProductType'
  }
}

var collection = Marcello["data"].Collection<Product, ProductIndex>("Products");
collection.Index(idx => idx.Name)
```

Accessing indexes has compiler support, and once the collection is created the syntax to access the index is quite simple.
The methods in the index are used to reference the specific index and are used to get the value to index form the object.

However, the index class itself has to follow the convention of exposing public methods which accept instance of the type of the collection (Product in this case) as single parameter.

**Comment by markmeeus** [comment]
To read data from the index, something like this:

``` c#
var typeIndex = collection.Index(idx => idx.Type);

foreach(var product in typeIndex.All){
  //products ordered by Type
}

foreach(var product in typeIndex.Descending){
  //products ordered by Type in descending order
}

foreach(var product in typeIndex.Greater(10)){
  //products starting from Type 10
}

foreach(var product in typeIndex.Lesser(10)){
  //products before type 10
}

foreach(var product in typeIndex.Between(5, 10)){
  //products with type 6, 7, 8, 9
}

foreach(var product in typeIndex.BetweenInclusive(5, 10)){
  //products with type 5, 6, 7, 8, 9, 10
}

foreach(var product in typeIndex.Where(20)){
  //products with Type 20
}

foreach(var product in typeIndex.Where(type => type > 10, type => type <= 14}){
  //Type 11, 12, 13, 14
}

foreach(var product in typeIndex.In(20, 30, 40){
  //products with Type 20
}

```

**Comment by markmeeus** [comment]
lower level

``` c#
var nameIndex = collection.Index(idx => idx.Name);
//nameIndex acts like some sort of cursor
var product = nameIndex.Current; //starts at first object if any

while(product != null){
  nameIndex.Next():
  var product = nameIndex.Current;
}

//index navigation methods
nameIndex.MoveFirst();
nameIndex.Next();
nameIndex.FindFirst(name => name.Compare("Toothbrush") >= 0);
```

All of the method in the comment above can be implemented with these 3

**Comment by markmeeus** [comment]
if we do this:

``` c#
class ProductIndex : MarcelloDB.IndexDefinition<Product>{
  public string Name(Product p){
   return p.Name
  }
}
```

Then the ID index definition can be implemented in the base class .
This will allow us to refactor the Find method into a simple call to the ID Index:

``` c#
T Find(object id){
  IndexDefinition.Index((def)=>def.ID).Find(id);
}
```

**Comment by markmeeus** [comment]
In theorie we could also use properties to define indexes

``` c#
class ProductIndex : MarcelloDB.IndexDefinition<Product>{
  public string Name {get;set;}
}
```

Which means we use the property with the same name (Name in this case) from Product.

**Comment by markmeeus** [comment]
Implemented in 0.2
